subscriber.go

```
@@ -65,40 +78,64 @@ func (s *subscriber) loop() {
 			// If there was already data, next is a closed channel.
 			// otherwise it is nil so won't pass through.
 		}
-		call, empty := s.popOne()
+		message, empty := s.popOne()
 		if empty {
 			next = nil
 		} else {
 			next = s.closed
 		}
-		// call *should* never be nil as we should only be calling
+		// message *should* never be nil as we should only be calling
 		// popOne in the situations where there is actually something to pop.
-		if call != nil {
+		if message != nil {
+			call := message.callback
 			s.logger.Tracef("exec callback %p (%d) func %p", s, s.id, s.handler)
 			s.handler(call.topic, call.data)
 			call.done()
+
+			// Consumed exposes information about how long a given message
+			// has been on the subscriber pending list. We can use this
+			// information to workout how much backpressure is being exhorted
+			// on the system at large.
+			s.metrics.Consumed(s.id, s.clock.Now().Sub(message.now))
 		}
```

if you did get nil, should we be reporting some sort of 'juju should be debugging this' rather than silently ignoring it? The earlier message seems very strong for dropping the condition on the floor.

---

Agreed, I've now log out with `Errorf`, so we can investigate when it does happen. https://github.com/juju/pubsub/pull/13/commits/f89121c392d9ebdd95131a0175cdba2329382c43

---

If you can do a quick sanity test once this lands, to make sure there isn't some unexpected condition where this is happening a lot. (do a bootstrap, deploy a couple things, make sure you don't see the new error message in the logs)